source/extensions/filters/http/ext_proc/processor_state.cc

```diff
@@ -272,111 +272,180 @@ absl::Status ProcessorState::handleHeadersResponse(const HeadersResponse& respon
-          }
-        }
-        ENVOY_LOG(debug, "Applying body response to buffered data. State = {}",
-                  static_cast<int>(callback_state_));
-        modifyBufferedData([&common_response](Buffer::Instance& data) {
-          MutationUtils::applyBodyMutations(common_response.body_mutation(), data);
-        });
-      }
-      clearWatermark();
-      onFinishProcessorCall(Grpc::Status::Ok);
-      should_continue = true;
-    } else if (callback_state_ == CallbackState::StreamedBodyCallback) {
-      absl::StatusOr<bool> result = handleBodyInStreamedState(common_response);
-      if (!result.ok()) {
-        return result.status();
-      }
-      should_continue = *result;
-    } else if (callback_state_ == CallbackState::BufferedPartialBodyCallback) {
-      // Apply changes to the buffer that we sent to the server
-      Buffer::OwnedImpl chunk_data;
-      auto chunk = dequeueStreamingChunk(chunk_data);
-      ENVOY_BUG(chunk != nullptr, "Bad partial body callback state");
-      if (common_response.has_header_mutation()) {
-        if (headers_ != nullptr) {
-          const auto mut_status = processHeaderMutation(common_response);
-          if (!mut_status.ok()) {
-            return mut_status;
-          }
-        } else {
-          ENVOY_LOG(debug, "Response had header mutations but headers aren't available");
-        }
-      }
-      if (common_response.has_body_mutation()) {
-        MutationUtils::applyBodyMutations(common_response.body_mutation(), chunk_data);
-      }
-      if (chunk_data.length() > 0) {
-        ENVOY_LOG(trace, "Injecting {} bytes of processed data to filter stream",
-                  chunk_data.length());
-        injectDataToFilterChain(chunk_data, chunk->end_stream);
-      }
-      should_continue = true;
-      clearWatermark();
-      onFinishProcessorCall(Grpc::Status::Ok);
-      partial_body_processed_ = true;
-
-      // If anything else is left on the queue, inject it too
-      if (chunkQueue().receivedData().length() > 0) {
-        const auto& all_data = consolidateStreamedChunks();
-        ENVOY_LOG(trace, "Injecting {} bytes of leftover data to filter stream",
-                  chunkQueue().receivedData().length());
-        injectDataToFilterChain(chunkQueue().receivedData(), all_data.end_stream);
-      }
-    } else {
-      // Fake a grpc error when processor state and received message type doesn't match, beware this
-      // is not an error from grpc.
-      onFinishProcessorCall(Grpc::Status::FailedPrecondition);
+  absl::StatusOr<bool> result;
+  switch (callback_state_) {
+  case CallbackState::BufferedBodyCallback:
+    result = handleBufferedBodyCallback(common_response);
+    break;
+  case CallbackState::StreamedBodyCallback:
+    result = handleStreamedBodyCallback(common_response);
+    break;
+  case CallbackState::BufferedPartialBodyCallback:
+    result = handleBufferedPartialBodyCallback(common_response);
+    break;
+  default:
```

Comment:
This seems to be dead code based on the context.  We can clean it up in the next PR.

Reply:
what happens if you remove the 'default' block here? Will it compile? if CallbackState is a C++ enum and you've covered all the options it's better to omit the 'default'. If there are cases which exist in the enum but you think they can't happen, we can use IS_ENVOY_BUG in the default to log it in production or crash in tests.

Reply:
Line 276 clearly checked the callback_state has to be one of these three body states. Thus, the "default" case will never be hit.  Let's just remove it then.

Reply:
@jmarantz It looks like `CallbackState` have more states than what we have in our ExtProc handling today,
```
enum class CallbackState {
    // Not waiting for anything
    Idle,
    // Waiting for a "headers" response. This may be true even if we started
    // to receive data frames for a message.
    HeadersCallback,
    // Waiting for a "body" response in buffered mode.
    BufferedBodyCallback,
    // Waiting for a "body" response in streaming mode.
    StreamedBodyCallback,
    // Waiting for a body callback in "buffered partial" mode.
    BufferedPartialBodyCallback,
    // Waiting for a "trailers" response.
    TrailersCallback,
};
```

When I try to remove `default` and add `IS_ENVOY_BUG(...)`, I see this error:

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
source/extensions/filters/http/ext_proc/processor_state.cc:284:11: error: enumeration values 'Idle', 'HeadersCallback', and 'TrailersCallback' not handled in switch [-Werror,-Wswitch]
  switch (callback_state_) {
          ^~~~~~~~~~~~~~~
1 error generated.
```

Reply:
Had you tried this:
default:
   IS_ENVOY_BUG("callback state is wrong");
   break;

Reply:
@yanjunxiang-google Okay, added that.

Reply:
I'm OK with this as is but it would be slightly better (IMO) to put the IS_ENVOY_BUG call first. No big deal.

I'd also slightly favor enumerating all the missing cases rather than use 'default'. That way if a new enum choice is added in the future, the compiler will point out the missing new case here, and the author of this code will then consider the new case for this block with it fresh in his/her mind.